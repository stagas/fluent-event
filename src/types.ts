export type Fn<P extends any[], R> = (...args: P) => R
