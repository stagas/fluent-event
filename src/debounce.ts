import { FluentFlags } from 'fluent-flags'
import type { Fn } from './types'

/**
 * Decorate function `fn` with debounce delay `ms`.
 *
 * Flags: `first` => run `fn` first, then debounce
 * ```ts
 * fn = (x: number) => console.log(x)
 * // => runs `fn` after 100ms following last call
 * cb = debounce()(fn, 100)
 * cb(1)
 * cb(2) // <- cb(2) wins, prints `2`
 *
 * // => runs `fn` first then debounces until 100ms of inactivity
 * cb = debounce().first(fn, 100)
 * cb(1) // <- cb(1) wins, prints `1`
 * cb(2)
 * ```
 */
export const debounce = () =>
  FluentFlags<'first'>(flags =>
    <P extends any[]>(fn: Fn<P, any>, ms: number) => {
      let id: any
      let runs = false

      const cb = flags.first
        ? (fn: any) => {
          clearTimeout(id)
          id = setTimeout(() => (runs = false), ms)
          if (runs) return
          runs = true
          fn()
        }
        : (fn: any) => {
          clearTimeout(id)
          id = setTimeout(fn, ms)
        }

      return function(this: any, ...args: P) {
        cb(() => fn.apply(this, args))
      }
    }
  )
