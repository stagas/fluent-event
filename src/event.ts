import { FluentFlags } from 'fluent-flags'

export type EventListener<E extends Event, T = HTMLElement | SVGElement | Window> = (
  this: T,
  event: E,
) => any

/**
 * Decorates event listener `fn`.
 *
 * Flags:
 * - `prevent` => `event.preventDefault()`
 * - `stop` => `event.stopPropagation()`
 * - `stop.immediate` => `event.stopImmediatePropagation()`
 *
 * ```ts
 * btn.onclick = event()(fn)
 * btn.onclick = event().prevent(fn)
 * btn.onclick = event().prevent.stop(fn)
 * btn.onclick = event().stop.immediate(fn)
 * ```
 */
export const event = () =>
  FluentFlags<'prevent' | 'stop' | 'immediate'>(flags =>
    <
      E extends Event,
      T extends HTMLElement | SVGElement | Window | GlobalEventHandlers,
    >(fn: EventListener<E, T> = () => {}): EventListener<E, T> =>
      function(this: T, event: E) {
        if (flags.prevent) event.preventDefault()
        if (flags.stop) flags.immediate ? event.stopImmediatePropagation() : event.stopPropagation()
        return fn.call(this, event)
      }
  )
