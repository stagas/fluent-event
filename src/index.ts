export * from './debounce'
export * from './event'
export * from './on'
export * from './queue'
export * from './types'
