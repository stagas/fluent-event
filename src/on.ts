import { FluentFlags } from 'fluent-flags'

function onEvent<T extends HTMLElement | SVGElement | Window, K extends keyof WindowEventMap>(
  el: T,
  type: K,
  listener?: (ev: WindowEventMap[K]) => any,
  options?: boolean | AddEventListenerOptions,
): () => void
function onEvent<T extends SVGElement, K extends keyof SVGElementEventMap>(
  el: T,
  type: K,
  listener?: (ev: SVGElementEventMap[K]) => any,
  options?: boolean | AddEventListenerOptions,
): () => void
function onEvent<T extends HTMLElement, K extends keyof HTMLElementEventMap>(
  el: T,
  type: K,
  listener?: (ev: HTMLElementEventMap[K]) => any,
  options?: boolean | AddEventListenerOptions,
) {
  el.addEventListener(type, listener!, options)
  return () => off(el, type, listener!, options)
}

/**
 * Adds a DOM event `listener` to an event `type` using `options` and returns its remover.
 *
 * Flags: `active` | `capture` | `once` | `passive`
 *
 * ```ts
 * on()(btn, 'click', fn)
 * on().once(btn, 'click', fn)
 * on().passive(div, 'wheel', fn)
 *
 * const off = on().passive.capture(btn, 'wheel', fn)
 * // ...later...
 * off() // remove listener
 * ```
 */
export const on = () =>
  FluentFlags<'active' | 'capture' | 'once' | 'passive'>(flags =>
    ((el: any, type: any, fn: any, options: any = {}) =>
      onEvent(el, type, fn, {
        ...(flags.active ? { passive: false } : null),
        ...flags,
        ...options,
      })) as typeof onEvent
  )

/**
 * Removes an event `listener` of type `type` from `el` using `options`.
 */
export const off = <T extends HTMLElement, K extends keyof HTMLElementEventMap>(
  el: T,
  type: K,
  listener: (ev: HTMLElementEventMap[K]) => any,
  options?: boolean | AddEventListenerOptions,
) => el.removeEventListener(type, listener, options)
