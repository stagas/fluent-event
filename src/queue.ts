import type { Fn } from './types'

const q =
  (queueFn: Fn<[Fn<any[], any>], any>) =>
  <P extends any[]>(fn: Fn<P, any>): Fn<P, void> => {
    let args: P | null
    let after: P | null
    const cb = function (this: any) {
      // and if there are more calls, schedule early
      if (after) queueFn(cb)

      // callback runs with first args
      fn.apply(this, args!)

      args = after ?? null
      after = null
    }
    return function (this: any, ...newArgs: P) {
      // initial call gets first args
      if (!args) {
        queueFn(cb)
        args = newArgs
      } else {
        // later calls get latest args
        after = newArgs
      }
    }
  }

/**
 * Queue.
 *
 * All queue functions are also throttled to once per invocation.
 */
export const queue = {
  /**
   * Decorate function with `requestAnimationFrame`.
   * ```ts
   * queue.raf(cb)
   * ```
   */
  raf: q(requestAnimationFrame),
  /**
   * Decorate function with `setTimeout`.
   * ```ts
   * queue.time(cb)
   * ```
   */
  time: q(setTimeout),
  /**
   * Decorate function with `queueMicrotask`.
   * ```ts
   * queue.task(cb)
   * ```
   */
  task: q(queueMicrotask),
}
