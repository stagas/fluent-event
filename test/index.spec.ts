import { debounce, event, on, queue } from '../src'

describe('all', () => {
  describe('event(fn)', () => {
    it('decorates listener', () => {
      let count = 0
      const btn = document.createElement('button')
      btn.onclick = event(() => count++)
      btn.click()
      expect(count).toEqual(1)
    })

    it('prevent', () => {
      let count = 0
      let prevent = 0
      let stop = 0
      const mock = {
        preventDefault() {
          prevent++
        },
        stopPropagation() {
          stop++
        },
      } as Event
      event.prevent(() => count++).call(this as any, mock)
      expect(count).toEqual(1)
      expect(prevent).toEqual(1)
      expect(stop).toEqual(0)
    })

    it('stop', () => {
      let count = 0
      let prevent = 0
      let stop = 0
      const mock = {
        preventDefault() {
          prevent++
        },
        stopPropagation() {
          stop++
        },
      } as Event
      event.stop(() => count++).call(this as any, mock)
      expect(count).toEqual(1)
      expect(prevent).toEqual(0)
      expect(stop).toEqual(1)
    })

    it('stop.immediate', () => {
      let count = 0
      let prevent = 0
      let stop = 0
      const mock = {
        preventDefault() {
          prevent++
        },
        stopImmediatePropagation() {
          stop++
        },
      } as Event
      event.stop.immediate(() => count++).call(this as any, mock)
      expect(count).toEqual(1)
      expect(prevent).toEqual(0)
      expect(stop).toEqual(1)
    })

    it('prevent.stop', () => {
      let count = 0
      let prevent = 0
      let stop = 0
      const mock = {
        preventDefault() {
          prevent++
        },
        stopPropagation() {
          stop++
        },
      } as Event
      event.prevent.stop(() => count++).call(this as any, mock)
      expect(count).toEqual(1)
      expect(prevent).toEqual(1)
      expect(stop).toEqual(1)
    })
  })

  describe('queue(fn)', () => {
    it('decorates function', async () => {
      let count = 0
      const cb = queue.task(() => count++)
      cb()
      expect(count).toEqual(0)
      await Promise.resolve()
      expect(count).toEqual(1)
    })

    it('raf', async () => {
      let count = 0
      const cb = queue.raf(() => count++)
      cb()
      expect(count).toEqual(0)
      await new Promise(resolve => requestAnimationFrame(resolve))
      expect(count).toEqual(1)
    })

    it('raf multiple times', async () => {
      let count = 0
      const cb = queue.raf(() => count++)
      cb()
      cb()
      cb()
      cb()
      expect(count).toEqual(0)
      const results: any = []
      await new Promise(resolve => {
        requestAnimationFrame(() => {
          results.push(count)
          requestAnimationFrame(resolve)
        })
      })
      results.push(count)
      expect(count).toEqual(2)
      expect(results).toMatchSnapshot()
    })

    it('raf last first and last args win', async () => {
      let value: any
      const cb = queue.raf(x => {
        value = x
      })
      cb(1)
      cb(2)
      cb(3)
      cb(4)
      expect(value).toBeUndefined()
      const results: any = []
      await new Promise(resolve => {
        requestAnimationFrame(() => {
          results.push(value)
          requestAnimationFrame(resolve)
        })
      })
      results.push(value)
      expect(value).toBe(4)
      expect(results).toMatchSnapshot()
    })

    it('time', async () => {
      let count = 0
      const cb = queue.time(() => count++)
      cb()
      expect(count).toEqual(0)
      await new Promise(resolve => setTimeout(resolve))
      expect(count).toEqual(1)
    })
  })

  describe('on/off', () => {
    it('adds event listener', () => {
      let count = 0
      const btn = document.createElement('button')
      const cb = () => count++
      const off = on(btn, 'click', cb)
      expect(count).toBe(0)
      btn.click()
      expect(count).toBe(1)
      btn.click()
      expect(count).toBe(2)
      off()
      btn.click()
      expect(count).toBe(2)
    })

    it('once', () => {
      let count = 0
      const btn = document.createElement('button')
      const cb = () => count++
      on.once(btn, 'click', cb)
      expect(count).toBe(0)
      btn.click()
      expect(count).toBe(1)
      btn.click()
      expect(count).toBe(1)
    })
  })

  describe('debounce', () => {
    it('debounces fn', async () => {
      let count = 0
      const cb = debounce(() => count++, 5)
      cb()
      cb()
      cb()
      expect(count).toBe(0)
      await new Promise(resolve => setTimeout(resolve, 10))
      expect(count).toBe(1)
    })

    it('debounces first run', async () => {
      let count = 0
      const cb = debounce.first(() => count++, 5)
      cb()
      cb()
      cb()
      expect(count).toBe(1)
      await new Promise(resolve => setTimeout(resolve, 10))
      expect(count).toBe(1)
      cb()
      expect(count).toBe(2)
    })

    it('passes arguments first', async () => {
      let count = 0
      const cb = debounce.first((x: number) => (count += x), 5)
      cb(2)
      cb(2)
      cb(2)
      expect(count).toBe(2)
      await new Promise(resolve => setTimeout(resolve, 10))
      expect(count).toBe(2)
      cb(2)
      expect(count).toBe(4)
    })

    it('passes arguments last win', async () => {
      let count = 0
      const cb = debounce((x: number) => (count += x), 5)
      cb(2)
      cb(2)
      cb(3)
      expect(count).toBe(0)
      await new Promise(resolve => setTimeout(resolve, 10))
      expect(count).toBe(3)
    })
  })
})
