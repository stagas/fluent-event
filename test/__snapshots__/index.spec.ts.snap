// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all queue(fn) raf last first and last args win 1`] = `
Array [
  1,
  4,
]
`;

exports[`all queue(fn) raf multiple times 1`] = `
Array [
  1,
  2,
]
`;
